### ISSUE TITLE
Reuse Bitmap, Paint & Canvas instances in Android LocalGlyphRasterizer

### ISSUE THREAD

**Issue by themics** [mapbox/mapbox-gl-native#12421]
**Mapbox SDK versions: 6.3.0-beta1** 

Since `LocalGlyphRasterizer.drawGlyphBitmap()` only called in a GL thread, couldn't we store these instances(per GL thread, perhaps `ThreadLocal`?) and reuse?

In this case of course we should clear the canvas by `drawColor(Color.TRANSPARENT, PorterDuff.Mode.CLEAR)` before drawing, but even with that, reusing is much more fast.

I tested a bit by rasterizing 1,000 korean characters with my Pixel 2, here's the result:
- Current Implementation
  - 1st attempt: 158ms
  - 2nd~ attempt with same characters: 32ms(world layout cached by Android system)
- Reusing implementation
  - 1st: 102ms (1.5x faster)
  - 2nd~: 10ms (3x faster)

**Comment by ChrisLoer** [comment]
Hi @themics, that sounds like a good suggestion! You're right, `drawGlyphBitmap()` should only be called from the "renderer" thread. I implemented `LocalGlyphRasterizer`, but I'm not normally a Java developer -- would you be interested in submitting a PR with your suggested changes?

Also I'm curious, did the glyph rasterization cost show up as a significant expense in part of your profiling? A 1.5x-3x speed-up is great, although I had the assumption that even on maps with lots of CJK characters, the rasterization cost would be a pretty small part of overall map loading time.

**Comment by ChrisLoer** [comment]
Ah I see, yes, it probably most makes sense for the re-used items to be members of `LocalGlyphRasterizer`, and then `drawGlyphBitmap` should become a member function instead of a static function. `android/src/text/local_glyph_rasterizer.cpp` will have to be updated to instantiate the java `LocalGlyphRasterizer` and hold onto it over the lifetime of `LocalGlyphRasterizer::Impl`. Are you familiar with JNI? I am no expert on JNI, but I recently used `jni::UniqueObject` to hold onto a long-lived JNI object in `android/src/text/collator.cpp`, you could try following that pattern.

Sounds like almost 9% improvement in time to render, or >100ms -- that's not _huge_, but it definitely sounds worth the effort. I'm really glad to hear you're getting use out of the local glyph generation feature!

If you got really ambitious and wanted to further optimize Android glyph generation, I suspect a good approach would be to batch up the drawing request so that all of the glyphs in a batch were drawn into a single bitmap, in order to minimize the JNI overhead.

**Comment by themics** [comment]
@ChrisLoer Yup turns out it was simple job, native `LocalGlyphRasterizer` was already in 1:1 relation with a renderer. I'll send a PR soon. Batching sounds really good solution - but I'm out of ambition for now(haha). Maybe I can take a look next time.
